pyoozie/tags.py

```diff
@@ -0,0 +1,313 @@
+# Copyright (c) 2017 "Shopify inc." All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
+from __future__ import unicode_literals
+from abc import ABCMeta, abstractmethod
+import re
+import yattag
+
+
+MAX_IDENTIFIER_LENGTH = 39
+REGEX_IDENTIFIER = r'^[a-zA-Z_][\-_a-zA-Z0-9]{0,38}$'
+COMPILED_REGEX_IDENTIFIER = re.compile(REGEX_IDENTIFIER)
+
+
+class IdentifierTooLongError(AssertionError):
+
+    def __init__(self, identifier):
+        AssertionError.__init__(self, "Identifier must be less than {max_length} chars long, '{identifier}' is "
+                                "{length}".format(max_length=MAX_IDENTIFIER_LENGTH,
+                                                  identifier=identifier,
+                                                  length=len(identifier)))
+        self.length = len(identifier)
+
+
+def _validate(identifier):
+
+    if len(identifier) > MAX_IDENTIFIER_LENGTH:
+        raise IdentifierTooLongError(identifier)
+
+    assert COMPILED_REGEX_IDENTIFIER.match(identifier), \
+        "Identifier must match {regex}, '{identifier}' does not".format(
+            regex=REGEX_IDENTIFIER,
+            identifier=identifier)
+
+    return identifier
+
+
+class Xml(object):
+    """An abstract object that can be serialized to XML."""
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self, xml_tag):
+        self.xml_tag = xml_tag
+
+    def xml(self, indent=False):
+        doc, tag, text = yattag.Doc().tagtext()
+        doc.asis("<?xml version='1.0' encoding='UTF-8'?>")
+        xml = self._xml(doc, tag, text).getvalue()
+        if indent:
+            return yattag.indent(xml, indentation=' ' * 4, newline='\r\n')
+        else:
+            return xml
+
+    @abstractmethod
+    def _xml(self, doc, tag, text):
+        raise NotImplementedError
+
+    def __str__(self):
+        return self.xml_tag
+
+
+class _PropertyList(Xml, dict):
+    """
+    Object used to represent Oozie workflow/coordinator property-value sets.
+
+    Generates XML of the form:
+    ...
+    <xml_tag>
+      <property>
+        <name>[PROPERTY-NAME]</name>
+        <value>[PROPERTY-VALUE]</value>
+      </property>
+      ...
+    </xml_tag>
+    """
+
+    def __init__(self, xml_tag, attributes=None, values=None):
+        Xml.__init__(self, xml_tag)
+        if values:
+            dict.__init__(self, values)
+        else:
+            dict.__init__(self)
+        self.attributes = attributes or {}
+
+    def _xml(self, doc, tag, text):
+        with tag(self.xml_tag, **self.attributes):
+            for name, value in sorted(self.items()):
+                with tag('property'):
+                    with tag('name'):
+                        doc.text('%s' % name)
+                    with tag('value'):
+                        doc.text('%s' % value if value is not None else '')
```

Comment:
Correct; I want to be able to serialize the boolean `False` as the string `False`.